src/sync/PostSyncProcessor.js

```diff
@@ -0,0 +1,164 @@
+/**
+ * All utility functions for processing synced records after being synced. These should changes
+ * should be detected by SyncQueue and sent back to the server on next sync.
+ */
+
+import { getNextNumber, NUMBER_SEQUENCE_KEYS } from '../database/utilities';
+
+const { REQUISITION_SERIAL_NUMBER, SUPPLIER_INVOICE_NUMBER } = NUMBER_SEQUENCE_KEYS;
+
+export class PostSyncProcessor {
+  constructor(database) {
+    this.database = database;
+    this.recordQueue = new Map(); // Map of [recordId, recordType]
+    this.actionQueue = [];
+    this.checkTables = this.checkTables.bind(this);
+    this.processRecordQueue = this.processRecordQueue.bind(this);
+    this.onDatabaseEvent = this.onDatabaseEvent.bind(this);
+    this.runActionQueue = this.runActionQueue.bind(this);
+    this.delegateByRecordType = this.delegateByRecordType.bind(this);
+    this.postProcessRequisition = this.postProcessRequisition.bind(this);
+    this.postProcessTransaction = this.postProcessTransaction.bind(this);
+
+    this.database.addListener(this.onDatabaseEvent);
+  }
+
+  /**
+ * Respond to a database change event. Must be called from within a database
+ * write transaction.
+ * @param  {string} changeType  The type of database change, e.g. CREATE, UPDATE, DELETE
+ * @param  {string} recordType  The type of record changed (from database schema)
+ * @param  {object} record      The record changed
+ * @param  {string} causedBy    The cause of this database event, either 'sync' or undefined
+ * @return {none}
+ * Runs all the post sync functions triggered by sync events on database through
+ * this.onDatabaseEvent
+ */
+  onDatabaseEvent(changeType, recordType, record, causedBy) {
+    console.log('event: ', causedBy, recordType);
+    // Exit if not a change caused by incoming sync
+    if (causedBy !== 'sync' || recordType === 'SyncOut') return;
+    if (this.recordQueue.has(record.id)) {
+      // Check if already in queue, remove old
+      this.recordQueue.delete(record.id);
+    }
+    // Add new entry at end of Map
+    this.recordQueue.set(record.id, recordType);
+  }
+
+  /**
+   * Runs post checks across tables, ensuring data is correct.
+   * Tables manually added as to not iterate over tables that don't have any post processing.
+   */
+  checkTables() {
```

Comment:
I think this method name could be a bit more descriptive. Like `processAnyFoos` where `Foo` is a word that means unprocessed record.